File: app/src/main/java/com/stupidtree/hichat/data/repository/ChatRepository.java
package com.stupidtree.hichat.data.repository;

import android.content.Context;
import android.content.IntentFilter;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.Transformations;

import com.stupidtree.hichat.data.AppDatabase;
import com.stupidtree.hichat.data.model.ChatMessage;
import com.stupidtree.hichat.data.source.ChatMessageDao;
import com.stupidtree.hichat.data.source.ChatMessageWebSource;
import com.stupidtree.hichat.data.source.SocketWebSource;
import com.stupidtree.hichat.ui.base.DataState;
import com.stupidtree.hichat.ui.chat.FriendStateTrigger;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.Collections;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import top.zibin.luban.Luban;
import top.zibin.luban.OnCompressListener;

import static com.stupidtree.hichat.socket.SocketIOClientService.ACTION_FRIEND_STATE_CHANGED;
import static com.stupidtree.hichat.socket.SocketIOClientService.ACTION_MESSAGE_SENT;
import static com.stupidtree.hichat.socket.SocketIOClientService.ACTION_RECEIVE_MESSAGE;

/**
 * 层次：Repository
 * 消息记录仓库
 */
public class ChatRepository {
    //不采用单例模式
    public static ChatRepository getInstance(@NonNull Context context) {
        return new ChatRepository(context.getApplicationContext());
    }


    //数据源1：网络类型数据，消息记录的网络数据源
    ChatMessageWebSource chatMessageWebSource;
    //数据源2：和后台服务通信的Service
    SocketWebSource socketWebSource;
    //数据源3：本地数据库的数据源
    ChatMessageDao chatMessageDao;

    // 动态数据对象：列表状态
    MediatorLiveData<DataState<List<ChatMessage>>> listDataState = new MediatorLiveData<>();
    LiveData<List<ChatMessage>> localListState = null; //本地获取数据
    LiveData<DataState<List<ChatMessage>>> webListState = null; //网络获取数据


    public ChatRepository(@NonNull Context context) {
        chatMessageWebSource = ChatMessageWebSource.getInstance();
        socketWebSource = new SocketWebSource();
        chatMessageDao = AppDatabase.getDatabase(context).chatMessageDao();
    }


    public MediatorLiveData<DataState<List<ChatMessage>>> getListDataState() {
        listDataState.addSource(socketWebSource.getNewMessageState(), message -> {
            if (message != null) {
                listDataState.removeSource(localListState);
                listDataState.setValue(new DataState<>(Collections.singletonList(message)).setListAction(DataState.LIST_ACTION.APPEND));
                saveMessageAsync(message);
            }
        });
        listDataState.addSource(getMessageSentSate(), chatMessageDataState -> {
            listDataState.removeSource(localListState);
            ChatMessage sentMessage = chatMessageDataState.getData();
            saveMessageAsync(sentMessage);
        });
        return listDataState;
    }


    /**
     * 动作：拉取聊天记录
     *
     * @param token          令牌
     * @param conversationId 对话id
     * @param fromId         现有列表顶部的消息id
     * @param pageSize       分页大小
     * @param action         操作：全部刷新或在头部插入
     */
    public void ActionFetchMessages(@NonNull String token, @Nullable String conversationId, Long fromId, int pageSize, DataState.LIST_ACTION action) {
        listDataState.removeSource(localListState);
        Log.e("fromId", String.valueOf(fromId));
        if (fromId == null) {
            localListState = chatMessageDao.getMessages(conversationId, pageSize);
        } else {
            localListState = chatMessageDao.getMessages(conversationId, pageSize, fromId);
        }
        final boolean[] tried = {false};
        listDataState.addSource(localListState, chatMessages -> {
            Log.e("onChanged", chatMessages.size() + "-" + tried[0] + "-" + action + "-" + fromId);
            if (!tried[0] && chatMessages.size() < pageSize) {
                listDataState.removeSource(webListState);
                webListState = chatMessageWebSource.getMessages(token, conversationId, String.valueOf(fromId), pageSize);
                listDataState.addSource(webListState, result -> {
                    tried[0] = true;
                    if (result.getState() == DataState.STATE.SUCCESS && result.getData().size() > 0) {
                        saveMessageAsync(result.getData());
                    } else {
                        listDataState.setValue(new DataState<>(chatMessages).setListAction(action));
                    }
                });
            } else if (!tried[0] && fromId == null && chatMessages.size() > 0) { //第一次获取，拉取新消息
                //先显示本地有的
                listDataState.setValue(new DataState<>(chatMessages).setListAction(action));
                //尝试查询本地缺的
                listDataState.removeSource(webListState);
                webListState = chatMessageWebSource.pullLatestMessages(token, conversationId, String.valueOf(chatMessages.get(0).getId()));
                listDataState.addSource(webListState, result -> {
                    tried[0] = true;
                    Log.e("补全本地消息", String.valueOf(result));
                    if (result.getState() == DataState.STATE.SUCCESS && result.getData().size() > 0) {
                        listDataState.removeSource(localListState);
                        saveMessageAsync(result.getData());
                        listDataState.setValue(new DataState<>(result.getData()).setListAction(DataState.LIST_ACTION.APPEND));
                    }
                });

            } else {
                listDataState.setValue(new DataState<>(chatMessages).setListAction(action));
            }
        });


    }

    /**
     * 动作：获取新消息
     *
     * @param token          令牌
     * @param conversationId 对话id
     * @param afterId        现有列表底部的消息id
     */
    public void ActionFetchNewMessages(@NonNull String token, @Nullable String conversationId, @NotNull Long afterId) {
        //尝试查询本地缺的
        listDataState.removeSource(webListState);
        webListState = chatMessageWebSource.pullLatestMessages(token, conversationId, String.valueOf(afterId));
        listDataState.addSource(webListState, result -> {
            Log.e("手动拉取本地未存新消息", afterId + "-" + result);
            if (result.getState() == DataState.STATE.SUCCESS && result.getData().size() > 0) {
                listDataState.removeSource(localListState);
                saveMessageAsync(result.getData());
                listDataState.setValue(new DataState<>(result.getData()).setListAction(DataState.LIST_ACTION.APPEND));
            }
        });
    }

    /**
     * 动作：发消息
     *
     * @param message 消息
     */
    public void ActionSendMessage(ChatMessage message) {
        socketWebSource.sendMessage(message);
        listDataState.setValue(new DataState<>(Collections.singletonList(message)).setListAction(DataState.LIST_ACTION.APPEND));
    }

    /**
     * 动作：标记对话全部已读
     *
     * @param context        上下文，为了发广播
     * @param userId         用户id
     * @param conversationId 对话id
     */
    public void ActionMarkAllRead(@NonNull Context context, @NonNull String userId, @NonNull String conversationId) {
        socketWebSource.markAllRead(context, userId, conversationId);
    }

    /**
     * 动作：标记某消息已读
     *
     * @param context        上下文
     * @param messageId      消息id
     * @param conversationId 对话id
     */
    public void ActionMarkRead(@NonNull Context context, @NonNull String messageId, @NonNull String conversationId) {
        socketWebSource.markRead(context, messageId, conversationId);
    }


    /**
     * 动作：进入对话
     *
     * @param context        上下文
     * @param userId         用户id
     * @param friendId       朋友id
     * @param conversationId 对话id
     */
    public void ActionGetIntoConversation(Context context, @NonNull String userId, @NonNull String friendId, @NonNull String conversationId) {
        socketWebSource.getIntoConversation(context, userId, friendId, conversationId);
    }

    /**
     * 动作：退出对话
     *
     * @param context        上下文
     * @param userId         用户id
     * @param conversationId 对话id
     */
    public void ActionLeftConversation(@NonNull Context context, @NonNull String userId, @NonNull String conversationId) {
        socketWebSource.leftConversation(context, userId, conversationId);
    }


    /**
     * 发送图片
     *
     * @param token    令牌
     * @param toId     朋友id
     * @param filePath 文件
     * @return 返回结果
     */
    public LiveData<DataState<ChatMessage>> ActionSendImageMessage(@NonNull Context context, @NonNull String token, @NonNull String fromId, @NonNull String toId, @NonNull String filePath) {
        MediatorLiveData<DataState<ChatMessage>> result = new MediatorLiveData<>();
        //读取图片文件
        File f = new File(filePath);
        ChatMessage tempMsg = new ChatMessage(fromId, toId, filePath);
        tempMsg.setType(ChatMessage.TYPE.IMG);
        listDataState.setValue(new DataState<>(Collections.singletonList(tempMsg)).setListAction(DataState.LIST_ACTION.APPEND));
        Luban.with(context)
                .setTargetDir(context.getExternalFilesDir("image").getAbsolutePath())
                .load(f)
                .setCompressListener(new OnCompressListener() { //设置回调
                    @Override
                    public void onStart() {
                        Log.e("luban", "开始压缩");
                    }

                    @Override
                    public void onSuccess(java.io.File file) {
                        Log.e("luban", "压缩成功");
                        // MutableLiveData<DataState<String>> result = new MutableLiveData<>();
                        RequestBody requestFile = RequestBody.create(MediaType.parse("multipart/form-data"), file);
                        //构造一个图片格式的POST表单
                        MultipartBody.Part body =
                                MultipartBody.Part.createFormData("upload", file.getName(), requestFile);
                        LiveData<DataState<ChatMessage>> sendResult = chatMessageWebSource.sendImageMessage(token, toId, body, tempMsg.getUuid());
                        result.addSource(sendResult, result::setValue);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e("luban", "压缩失败");
                    }
                }).launch();    //启动压缩

        return result;
    }


    public void bindService(Context context) {
        IntentFilter IF = new IntentFilter();
        IF.addAction(ACTION_RECEIVE_MESSAGE);
        IF.addAction(ACTION_MESSAGE_SENT);
        IF.addAction(ACTION_FRIEND_STATE_CHANGED);
        context.registerReceiver(socketWebSource, IF);
        socketWebSource.bindService("Chat", context);
    }

    public void unbindService(Context context) {
        context.unregisterReceiver(socketWebSource);
        socketWebSource.unbindService(context);
    }

    private void saveMessageAsync(ChatMessage chatMessage) {
        new Thread(() -> chatMessageDao.saveMessage(Collections.singletonList(chatMessage))).start();
    }

    private void saveMessageAsync(List<ChatMessage> chatMessages) {
        new Thread(() -> chatMessageDao.saveMessage(chatMessages)).start();
    }


    public MutableLiveData<DataState<ChatMessage>> getMessageSentSate() {
        return socketWebSource.getMessageSentSate();
    }


    public MutableLiveData<FriendStateTrigger> getFriendsStateController() {
        return socketWebSource.getFriendStateController();
    }


}

File: app/src/main/java/com/stupidtree/cloudliter/data/repository/ChatRepository.kt
package com.stupidtree.cloudliter.data.repository

import android.content.Context
import android.content.IntentFilter
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.stupidtree.cloudliter.data.AppDatabase
import com.stupidtree.cloudliter.data.model.ChatMessage
import com.stupidtree.cloudliter.data.model.Conversation
import com.stupidtree.cloudliter.data.source.dao.ChatMessageDao
import com.stupidtree.cloudliter.data.source.websource.ChatMessageWebSource
import com.stupidtree.cloudliter.data.source.websource.SocketWebSource
import com.stupidtree.cloudliter.service.socket.SocketIOClientService
import com.stupidtree.component.data.DataState
import com.stupidtree.component.data.DataState.LIST_ACTION
import com.stupidtree.cloudliter.ui.chat.FriendStateTrigger
import com.stupidtree.cloudliter.ui.chat.MessageReadNotification
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import top.zibin.luban.Luban
import top.zibin.luban.OnCompressListener
import java.io.File
import java.sql.Timestamp

/**
 * 层次：Repository
 * 消息记录仓库
 */
class ChatRepository(context: Context) {
    //数据源1：网络类型数据，消息记录的网络数据源
    var chatMessageWebSource: ChatMessageWebSource = ChatMessageWebSource.instance!!

    //数据源2：和后台服务通信的Service
    private var socketWebSource: SocketWebSource = SocketWebSource()

    //数据源3：本地数据库的数据源
    private var chatMessageDao: ChatMessageDao = AppDatabase.getDatabase(context).chatMessageDao()

    // 动态数据对象：列表状态
    private var listDataState = MediatorLiveData<DataState<List<ChatMessage>?>>()
    private var localListState: LiveData<List<ChatMessage>?>? = null //本地获取数据
    private var webListState: LiveData<DataState<List<ChatMessage>?>>? = null //网络获取数据
    fun getListDataState(): MediatorLiveData<DataState<List<ChatMessage>?>> {
        listDataState.addSource(socketWebSource.newMessageState) { message: ChatMessage? ->
            if (message != null) {
                listDataState.removeSource(localListState!!)
                listDataState.value = DataState(listOf(message) as List?).setListAction(LIST_ACTION.APPEND_ONE)
                saveMessageAsync(message)
            }
        }
        listDataState.addSource(messageReadState) { messageReadNotificationDataState: DataState<MessageReadNotification> ->
            listDataState.removeSource(localListState!!)
            markMessageReadAsync(messageReadNotificationDataState.data)
        }
        return listDataState
    }

    /**
     * 动作：拉取聊天记录
     *
     * @param token          令牌
     * @param conversationId 对话id
     * @param topId          现有列表顶部的消息id
     * @param pageSize       分页大小
     * @param action         操作：全部刷新或在头部插入
     */
    fun actionFetchMessages(token: String, conversationId: String, topId: String?, topTime: Timestamp?, pageSize: Int, action: LIST_ACTION) {
        localListState?.let { listDataState.removeSource(it) }
        localListState = if (topId == null) {
            chatMessageDao.getMessages(conversationId, pageSize)
        } else {
            chatMessageDao.getMessages(conversationId, pageSize, topTime)
        }
        listDataState.addSource(localListState!!) { chatMessages ->
            listDataState.value = DataState(chatMessages).setListAction(action)
            if (action === LIST_ACTION.REPLACE_ALL && topId == null && chatMessages != null && chatMessages.isNotEmpty()) { //第一次获取，且本地已有消息
                //那么从本地的最早消息开始，把消息全部拉取更新
                webListState?.let { listDataState.removeSource(it) }
                webListState = chatMessageWebSource.getMessagesAfter(token, conversationId, chatMessages[chatMessages.size - 1].id, true)
                listDataState.addSource(webListState!!) { result ->
                    if (result.state === DataState.STATE.SUCCESS && result.data!!.isNotEmpty()) {
                        listDataState.removeSource(localListState!!)
                        saveMessageAsync(result.data?.toList())
                        listDataState.value = result.setFromCache(false).setListAction(action)
                    }
                }
            } else { //本地无消息/上拉加载
                webListState?.let { listDataState.removeSource(it) }
                webListState = chatMessageWebSource.getMessages(token, conversationId, topId, pageSize)
                listDataState.addSource(webListState!!) { result ->
                    if (result.state === DataState.STATE.SUCCESS && result.data!!.isNotEmpty()) {
                        listDataState.removeSource(localListState!!)
                        saveMessageAsync(result.data?.toList())
                        if (chatMessages != null) {
                            listDataState.value = result.setListAction(action).setFromCache(!chatMessages.isNotEmpty())
                        }
                    }
                }
            }
        }
    }


    /**
     * 动作：获取新消息
     *
     * @param token          令牌
     * @param conversationId 对话id
     * @param afterId        现有列表底部的消息id
     */
    fun actionFetchNewMessages(token: String, conversationId: String, afterId: String?) {
        webListState?.let { listDataState.removeSource(it) }
        webListState = chatMessageWebSource.getMessagesAfter(token, conversationId, afterId, false)
        listDataState.addSource(webListState!!) { result ->
            Log.e("手动拉取本地未存新消息", "$afterId-$result")
            if (result.state === DataState.STATE.SUCCESS && result.data!!.isNotEmpty()) {
                listDataState.removeSource(localListState!!)
                saveMessageAsync(ArrayList(result.data!!))
                listDataState.value = DataState(result.data).setListAction(LIST_ACTION.APPEND)
            }
        }
    }

    /**
     * 动作：标记对话全部已读
     *
     * @param context        上下文，为了发广播
     * @param userId         用户id
     * @param conversationId 对话id
     */
    fun actionMarkAllRead(context: Context, type: Conversation.TYPE, userId: String, conversationId: String, topTime: Timestamp, num: Int) {
        socketWebSource.markAllRead(context, type, userId, conversationId, topTime, num)
    }

    /**
     * 动作：标记某消息已读
     *
     * @param context        上下文
     * @param messageId      消息id
     * @param conversationId 对话id
     */
    fun actionMarkRead(context: Context, type: Conversation.TYPE, userId: String, messageId: String, conversationId: String) {
        socketWebSource.markRead(context, type, userId, messageId, conversationId)
    }

    /**
     * 动作：进入对话
     *
     * @param context        上下文
     * @param userId         用户id
     */
    fun actionGetIntoConversation(context: Context, userId: String, conversationId: String) {
        socketWebSource.getIntoConversation(context, userId, conversationId)
    }

    /**
     * 动作：退出对话
     *
     * @param context        上下文
     * @param userId         用户id
     * @param conversationId 对话id
     */
    fun actionLeftConversation(context: Context, userId: String, conversationId: String) {
        socketWebSource.leftConversation(context, userId, conversationId)
    }


    fun sendTextMessage(token: String, fromId: String, conversationId: String, content: String): LiveData<Pair<DataState<ChatMessage?>, String>> {
        val message = ChatMessage(fromId, content, conversationId)
        listDataState.value = DataState(listOf(message) as List?).setListAction(LIST_ACTION.APPEND_ONE)
        return Transformations.switchMap(chatMessageWebSource.sendTextMessage(token, fromId, conversationId, content, message.uuid)) {
            return@switchMap MutableLiveData(Pair(it, message.uuid))
        }
    }

    /**
     * 发送图片
     *
     * @param token    令牌
     * @param toId     朋友id
     * @param filePath 文件
     * @return 返回结果
     */
    fun sendImageMessage(context: Context, token: String, fromId: String, conversationId: String, filePath: String): LiveData<Pair<DataState<ChatMessage?>, String>> {
        val result = MediatorLiveData<Pair<DataState<ChatMessage?>, String>>()
        //读取图片文件
        val f = File(filePath)
        val tempMsg = ChatMessage(fromId, filePath, conversationId)
        tempMsg.setType(ChatMessage.TYPE.IMG)
        listDataState.value = DataState(listOf(tempMsg) as List?).setListAction(LIST_ACTION.APPEND_ONE)
        Luban.with(context)
                .setTargetDir(context.getExternalFilesDir("image")!!.absolutePath)
                .load(f)
                .setCompressListener(object : OnCompressListener {
                    //设置回调
                    override fun onStart() {
                        Log.e("luban", "开始压缩")
                    }

                    override fun onSuccess(file: File) {
                        Log.e("luban", "压缩成功")
                        // MutableLiveData<DataState<String>> result = new MutableLiveData<>();
                        val requestFile = RequestBody.create(MediaType.parse("multipart/form-data"), file)
                        //构造一个图片格式的POST表单
                        val body = MultipartBody.Part.createFormData("upload", file.name, requestFile)
                        val sendResult = chatMessageWebSource.sendImageMessage(token, conversationId, body, tempMsg.uuid)
                        result.addSource(sendResult) { v ->
                            result.value = Pair(v, tempMsg.uuid)
                        }
                    }

                    override fun onError(e: Throwable) {
                        Log.e("luban", "压缩失败")
                    }
                }).launch() //启动压缩
        return result
    }


    /**
     * 发送图片
     *
     * @param token    令牌
     * @param toId     朋友id
     * @param filePath 文件
     * @return 返回结果
     */
    fun actionSendVoiceMessage(token: String, fromId: String, conversationId: String, filePath: String, voiceSeconds: Int): LiveData<Pair<DataState<ChatMessage?>, String>> {
        val result = MediatorLiveData<Pair<DataState<ChatMessage?>, String>>()
        //读取图片文件
        val f = File(filePath)
        val tempMsg = ChatMessage(fromId, filePath, conversationId)
        tempMsg.setType(ChatMessage.TYPE.VOICE)
        tempMsg.extra = voiceSeconds.toString()
        listDataState.value = DataState(listOf(tempMsg) as List?).setListAction(LIST_ACTION.APPEND_ONE)
        val requestFile = RequestBody.create(MediaType.parse("multipart/form-data"), f)
        //构造一个图片格式的POST表单
        val body = MultipartBody.Part.createFormData("upload", f.name, requestFile)
        val sendResult = chatMessageWebSource.sendVoiceMessage(token, conversationId, body, tempMsg.uuid, voiceSeconds)
        result.addSource(sendResult) { v ->
            result.value = Pair(v, tempMsg.uuid)
        }
        return result
    }


    /**
     * 开始语音合成
     */
    fun startTTS(token: String, message: ChatMessage): LiveData<DataState<ChatMessage?>> {
        return Transformations.map(chatMessageWebSource.startTTS(token, message.id)) {
            if (it.state == DataState.STATE.SUCCESS) {
                it.data?.let { cm ->
                    cm.id = message.id
                    saveMessageAsync(cm)
                }
            }
            return@map it
        }
    }


    fun bindService(context: Context) {
        val IF = IntentFilter()
        IF.addAction(SocketIOClientService.RECEIVE_RECEIVE_MESSAGE)
        IF.addAction(SocketIOClientService.RECEIVE_FRIEND_STATE_CHANGED)
        IF.addAction(SocketIOClientService.RECEIVE_MESSAGE_READ)
        IF.addAction(SocketIOClientService.RECEIVE_UNREAD_MESSAGE)
        context.registerReceiver(socketWebSource, IF)
        //socketWebSource.bindService("Chat", context)
    }

    fun unbindService(context: Context) {
        context.unregisterReceiver(socketWebSource)
        // socketWebSource.unbindService(context)
    }

    fun saveMessageAsync(chatMessage: ChatMessage) {
        Thread { chatMessageDao.saveMessage(listOf(chatMessage)) }.start()
    }

    private fun markMessageReadAsync(notification: MessageReadNotification?) {
//        if (notification!!.type == MessageReadNotification.TYPE.ALL) {
        Thread {
            notification?.messageInfo?.let {
                for (x in it.entries) {
                    chatMessageDao.messageRead(x.key, x.value)
                }
            }

        }.start()
//        } else {
//            Thread { notification.id?.let { chatMessageDao.messageRead(it) } }.start()
//        }
    }

    private fun saveMessageAsync(chatMessages: List<ChatMessage>?) {
        Thread { chatMessages?.let { chatMessageDao.saveMessage(it.toList()) } }.start()
    }


    val onlineStateController: MutableLiveData<FriendStateTrigger>
        get() = socketWebSource.onlineStateController

    val messageReadState: MutableLiveData<DataState<MessageReadNotification>>
        get() = socketWebSource.messageReadState


    companion object {
        //不采用单例模式
        @JvmStatic
        fun getInstance(context: Context): ChatRepository {
            return ChatRepository(context.applicationContext)
        }
    }

}